File: backend/decryptor/eapi.py
# SPDX-FileCopyrightText: Copyright (c) 2024 沉默の金 <cmzj@cmzj.org>
# SPDX-License-Identifier: GPL-3.0-only

import binascii
import hashlib
import json
from base64 import b64decode, b64encode

from pyaes import AESModeOfOperationECB


def pkcs7_pad(data: bytes, block_size: int = 16) -> bytes:
    pad_len = block_size - (len(data) % block_size)
    padding = bytes([pad_len] * pad_len)
    return data + padding


def pkcs7_unpad(data: bytes) -> bytes:
    pad_len = data[-1]
    if pad_len < 1 or pad_len > 16:
        msg = "Invalid padding encountered."
        raise ValueError(msg)
    return data[:-pad_len]


def aes_encrypt(data: str | bytes, key: bytes) -> bytes:
    if isinstance(data, str):
        data = data.encode()
    padded_data = pkcs7_pad(data)  # Ensure the data is padded
    aes = AESModeOfOperationECB(key)  # Using ECB mode
    encrypted_data = b''
    for i in range(0, len(padded_data), 16):
        encrypted_data += aes.encrypt(padded_data[i:i + 16])  # Encrypt in 16-byte blocks
    return encrypted_data


def aes_decrypt(cipher_buffer: bytes, key: bytes) -> bytes:
    aes = AESModeOfOperationECB(key)  # Using ECB mode
    decrypted_data = b''
    for i in range(0, len(cipher_buffer), 16):
        decrypted_data += aes.decrypt(cipher_buffer[i:i + 16])  # Decrypt in 16-byte blocks
    return pkcs7_unpad(decrypted_data)  # Remove padding after decryption


def eapi_params_encrypt(path: bytes, params: dict) -> str:
    """eapi接口参数加密.

    :param path: url路径
    :param params: 明文参数
    :return str: 请求data.
    """
    params_bytes = json.dumps(params, separators=(',', ':')).encode()
    sign_src = b'nobody' + path + b'use' + params_bytes + b'md5forencrypt'
    sign = hashlib.md5(sign_src).hexdigest()  # noqa: S324
    aes_src = path + b'-36cd479b6b5-' + params_bytes + b'-36cd479b6b5-' + sign.encode()
    encrypted_data = aes_encrypt(aes_src, b'e82ckenh8dichen8')
    return f"params={binascii.hexlify(encrypted_data).upper().decode()}"


def eapi_params_decrypt(encrypted_text: str) -> dict:
    """解密使用 _eapi_encrypt 函数加密的文本.

    :param crypto_text: 加密文本
    :return: 解密后的 dict 对象
    """
    encrypted_bytes = binascii.unhexlify(encrypted_text)
    decrypted_data = aes_decrypt(encrypted_bytes, b'e82ckenh8dichen8')
    decrypted_text = decrypted_data.decode()
    parts = decrypted_text.split('-36cd479b6b5-')
    path = parts[0].encode()
    params = json.loads(parts[1])
    sign_src = b'nobody' + path + b'use' + json.dumps(params, separators=(',', ':')).encode() + b'md5forencrypt'
    m = hashlib.md5()  # noqa: S324
    m.update(sign_src)
    return params


def get_cache_key(data: str | bytes) -> str:
    return b64encode(aes_encrypt(data, b")(13daqP@ssw0rd~")).decode()


def cache_key_decrypt(data: str | bytes) -> str:
    return aes_decrypt(b64decode(data), b")(13daqP@ssw0rd~").decode()


def eapi_response_decrypt(cipher_buffer: bytes) -> bytes:
    return aes_decrypt(cipher_buffer, b'e82ckenh8dichen8')

File: packages/lddc_lyrics_core/lib/src/cryptography/eapi.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';
import 'package:pointycastle/export.dart' hide Digest;

const String _deviceIdXorKey = '3go8&\$8*3*3h0k(2)2';
const String _delimiterText = '-36cd479b6b5-';
final Uint8List _delimiter = Uint8List.fromList(ascii.encode(_delimiterText));
final Uint8List _nobody = Uint8List.fromList(ascii.encode('nobody'));
final Uint8List _use = Uint8List.fromList(ascii.encode('use'));
final Uint8List _md5ForEncrypt = Uint8List.fromList(
  ascii.encode('md5forencrypt'),
);
final Uint8List _eapiAesKey = Uint8List.fromList(
  ascii.encode('e82ckenh8dichen8'),
);
final Uint8List _cacheAesKey = Uint8List.fromList(
  ascii.encode(')(13daqP@ssw0rd~'),
);

/// 迁移自Python版 `eapi.py::eapi_params_encrypt`。
String eapiParamsEncrypt({
  required Uint8List path,
  required Map<String, Object?> params,
}) {
  final Uint8List paramsBytes = Uint8List.fromList(
    utf8.encode(jsonEncode(params)),
  );
  final Uint8List signSrc =
      (BytesBuilder(copy: false)
            ..add(_nobody)
            ..add(path)
            ..add(_use)
            ..add(paramsBytes)
            ..add(_md5ForEncrypt))
          .takeBytes();
  final String sign = md5.convert(signSrc).toString();
  final Uint8List aesSrc =
      (BytesBuilder(copy: false)
            ..add(path)
            ..add(_delimiter)
            ..add(paramsBytes)
            ..add(_delimiter)
            ..add(ascii.encode(sign)))
          .takeBytes();
  final Uint8List encryptedData = _aesEncrypt(aesSrc, _eapiAesKey);
  return 'params=${_encodeHex(encryptedData).toUpperCase()}';
}

/// 迁移自Python版 `eapi.py::eapi_params_decrypt`。
Map<String, Object?> eapiParamsDecrypt(String encryptedText) {
  final Uint8List encryptedBytes = _decodeHex(encryptedText);
  final Uint8List decryptedData = _aesDecrypt(encryptedBytes, _eapiAesKey);
  final String decryptedText = utf8.decode(decryptedData);
  final int firstDelimiter = decryptedText.indexOf(_delimiterText);
  final int lastDelimiter = decryptedText.lastIndexOf(_delimiterText);
  if (firstDelimiter < 0 || lastDelimiter <= firstDelimiter) {
    throw const FormatException('eapi_params_decrypt 数据格式非法');
  }

  final Uint8List path = Uint8List.fromList(
    utf8.encode(decryptedText.substring(0, firstDelimiter)),
  );
  final String paramsText = decryptedText.substring(
    firstDelimiter + _delimiterText.length,
    lastDelimiter,
  );
  final dynamic rawParams = jsonDecode(paramsText);
  if (rawParams is! Map) {
    throw const FormatException('eapi_params_decrypt params 不是对象');
  }
  final Map<String, Object?> params = rawParams.cast<String, Object?>();

  // 对齐Python版：重算 sign 但不做校验，仅保持等效执行路径。
  final Uint8List signSrc =
      (BytesBuilder(copy: false)
            ..add(_nobody)
            ..add(path)
            ..add(_use)
            ..add(utf8.encode(jsonEncode(params)))
            ..add(_md5ForEncrypt))
          .takeBytes();
  md5.convert(signSrc);
  return params;
}

/// 迁移自Python版 `eapi.py::get_cache_key`。
String eapiGetCacheKey(Uint8List data) {
  return base64Encode(_aesEncrypt(data, _cacheAesKey));
}

/// 迁移自Python版 `eapi.py::cache_key_decrypt`。
String eapiCacheKeyDecrypt(String data) {
  final Uint8List decrypted = _aesDecrypt(base64Decode(data), _cacheAesKey);
  return utf8.decode(decrypted);
}

/// 迁移自Python版 `eapi.py::eapi_response_decrypt`。
Uint8List eapiResponseDecrypt(Uint8List cipherBuffer) {
  return _aesDecrypt(cipherBuffer, _eapiAesKey);
}

/// 迁移自Python版 `eapi.py::get_anonimous_username`。
String eapiGetAnonymousUsername(String deviceId) {
  final StringBuffer xored = StringBuffer();
  final int keyLength = _deviceIdXorKey.length;
  for (int i = 0; i < deviceId.length; i++) {
    final int xorCode =
        deviceId.codeUnitAt(i) ^ _deviceIdXorKey.codeUnitAt(i % keyLength);
    xored.writeCharCode(xorCode);
  }
  final Digest md5Digest = md5.convert(utf8.encode(xored.toString()));
  final String combined = '$deviceId ${base64Encode(md5Digest.bytes)}';
  return base64Encode(utf8.encode(combined));
}

Uint8List _pkcs7Pad(Uint8List data, {int blockSize = 16}) {
  final int padLen = blockSize - (data.length % blockSize);
  return Uint8List(data.length + padLen)
    ..setRange(0, data.length, data)
    ..fillRange(data.length, data.length + padLen, padLen);
}

Uint8List _pkcs7Unpad(Uint8List data) {
  if (data.isEmpty) {
    throw const FormatException('PKCS#7 padding 数据为空');
  }
  if (data.length % 16 != 0) {
    throw FormatException('PKCS#7 padding 数据长度不是 16 的倍数: ${data.length}');
  }
  final int padLen = data[data.length - 1];
  if (padLen < 1 || padLen > 16) {
    throw const FormatException('Invalid padding encountered.');
  }
  if (padLen > data.length) {
    throw FormatException('PKCS#7 padding 长度超出数据长度: $padLen');
  }
  for (int index = data.length - padLen; index < data.length; index += 1) {
    if (data[index] != padLen) {
      throw FormatException(
        'PKCS#7 padding 字节不一致: index=$index expected=$padLen actual=${data[index]}',
      );
    }
  }
  return Uint8List.sublistView(data, 0, data.length - padLen);
}

Uint8List _aesEncrypt(Uint8List data, Uint8List key) {
  final Uint8List padded = _pkcs7Pad(data);
  return _aesEcbCrypt(padded, key, encrypt: true);
}

Uint8List _aesDecrypt(Uint8List cipherBuffer, Uint8List key) {
  final Uint8List decrypted = _aesEcbCrypt(cipherBuffer, key, encrypt: false);
  return _pkcs7Unpad(decrypted);
}

Uint8List _aesEcbCrypt(Uint8List data, Uint8List key, {required bool encrypt}) {
  if (data.length % 16 != 0) {
    throw ArgumentError.value(
      data.length,
      'data.length',
      'AES ECB 输入必须是 16 的倍数',
    );
  }
  final BlockCipher cipher = ECBBlockCipher(AESEngine())
    ..init(encrypt, KeyParameter(key));
  final Uint8List output = Uint8List(data.length);
  for (int i = 0; i < data.length; i += 16) {
    cipher.processBlock(data, i, output, i);
  }
  return output;
}

String _encodeHex(Uint8List data) {
  final StringBuffer buffer = StringBuffer();
  for (final int byte in data) {
    buffer.write(byte.toRadixString(16).padLeft(2, '0'));
  }
  return buffer.toString();
}

Uint8List _decodeHex(String hex) {
  final String normalized = hex.trim();
  if (normalized.length.isOdd) {
    throw FormatException('非法十六进制字符串长度: ${normalized.length}');
  }
  final Uint8List bytes = Uint8List(normalized.length ~/ 2);
  for (int i = 0; i < bytes.length; i++) {
    final int start = i * 2;
    final int? value = int.tryParse(
      normalized.substring(start, start + 2),
      radix: 16,
    );
    if (value == null) {
      throw FormatException(
        '非法十六进制内容: ${normalized.substring(start, start + 2)}',
      );
    }
    bytes[i] = value;
  }
  return bytes;
}
